# create pbs script to run shiver in individual IDs.
library(DescTools)

#outfiles <- list.files("/rds/general/user/fferre15/home/Ethics-HIV/small_pop/sim1/res_sim1", full.names = TRUE)
#outfiles <- list.files("output_deepseq1/vts/merged_trees/Illumina_reads/results_sampling/")
outfiles <- list.files("/Users/user/Desktop/teste/storage_test", full.names = TRUE)



#location of IDS
loc_IDs <- "output_deepseq/vts/merged_trees/Illumina_reads/results_sampling"


#count number of files
n_files <- length(outfiles)

for(i in 1:length(outfiles)){

  pbsfilename <- paste("shiver", SplitPath(outfiles[i])$filename, "pbs", sep = ".")

  ID_dirs <- paste(outfiles[i], loc_IDs, sep = "/")

  ID_count <- list.files(ID_dirs)

  if(length(ID_count) < 10000){

    array_number <- paste("#PBS", " -J", " 1-", length(ID_count), sep = "")

  }

  for(j in 1:length(ID_count)){

    pbstext <- paste("#PBS -l walltime=72:00:00",
                     "#PBS -l select=1:ncpus=2:mem=15gb",
                     "#PBS -o sim1_stage3.stdout",
                     "#PBS -e sim1_stage3.stderr",
                     array_number,
                     sep = "\n")

    pbstext <- paste(pbstext,
                     "\n",
                     "## load in the R environment",
                     "module load anaconda3/personal",
                     "source activate deep_seq_analysis",

                     "\n",
                     "## copy required files to the temporary directory on the compute node",
                     "WORKDIR=${EPHEMERAL}/${PBS_JOBID}",
                     "mkdir -p $WORKDIR",
                     "cd $WORKDIR",

                     "\n",
                     "#copy R scripts",
                     "cp $HOME/Ethics-HIV/small_pop/Rscripts/4.Map_reads_shiver_v2.R .",

                     "\n",
                     "#copy Softwares and other input files",
                     "#that we will need to run the abovementioned R scripts",
                     "cp $HOME/Ethics-HIV/small_pop/input_files/reference_subtypeB_completeGenome.fasta .",

                     "cp $HOME/Ethics-HIV/Programs/shiver.tar.gz .",

                     "\n",
                     "#untar shiver program",
                     "tar -xzvf shiver.tar.gz ",
                     sep = "\n")

    #copy files to run shiver
    #get main dir name
    main_dirName <- SplitPath(outfiles[i])$filename

    #create dirs to copy files
    new_dir <- paste(loc_IDs, ID_count[j], sep = "/")

    fileToAnalyse <- paste("cp -a $WORK/Ethics-HIV/small_pop/sim1/res_sim1",
                           main_dirName,
                           loc_IDs,
                           ID_count[j],
                           ".",
                           sep = "/")



    fileToAnalyse <- paste(fileToAnalyse, paste(new_dir, ".", sep = "/"), sep = " ")

    tar_filename <- paste("results_", main_dirName, ID_count[j], ".tar.gz", sep = "")

    pbstext <- paste(pbstext,


                "#make new dir",
                paste("mkdir", "-p", new_dir, sep = " "),

                "\n",
                "#cp network simulations from first round of simulations",
                fileToAnalyse,

                "\n",
                "## Run R scripts",
                "Rscript 4.Map_reads_shiver_v2.R",

                "\n",
                "# remove simulated Illumina reads to decrease storage spage",
                paste("rm ", new_dir, "/*.fq.gz", sep = ""),

                "\n",
                "#remove tmp files generated by shiver",
                paste("rm ", new_dir, "/shiver/temp_*", sep = ""),

                "\n",
                "#compress files",
                paste("tar -czvf",
                      tar_filename,
                      "./output_deepseq/vts/merged_trees/Illumina_reads/results_sampling",
                      sep = " "),

                "\n",
                "#Remove directories for shiver",
                "# so they do not get transferred back to home directory",
                "rm -rf shiver",
                "rm shiver.tar.gz",
                "rm -rf ./output_deepseq/vts/merged_trees/Illumina_reads/results_sampling",


                "## copy files back to the submission directory - anything not copied back will be lost",
                paste("cp -a $WORKDIR $WORK/Ethics-HIV/small_pop/sim1/res_sim1",
                      main_dirName,
                      loc_IDs,
                      paste("newID_$PBS_ARRAY_INDEX", "/", sep = ""),
                      sep = "/"),

                sep = "\n")

  }

  write.table(pbstext, file = pbsfilename, quote = FALSE,
              row.names = FALSE, col.names = FALSE)
}






